tools/bazar/services/EntryManager.php

```diff
@@ -454,6 +459,72 @@ public function update($tag, $data, $semantic = false, $replace = false)
         return $data;
     }
 
+    /**
+     * Replace the field values which are restricted at reading and writing. These values must be loaded to save them
+     * without user modification.
+     * As the fields are rectricted at reading, the right must be bypassed to load them.
+     *
+     * @param array $data the provided data to update
+     * @return array the data with the restricted values added
+     * @throws Exception
+     */
+    protected function assignRestrictedFields(array $data)
+    {
+        // not possible to init the formManager in the constructor because of circular reference problem
+        $form = $this->wiki->services->get(FormManager::class)->getOne($data['id_typeannonce']);
+
+        // check if there are some restricted fields at both reading and writing
+        $restrictedFields = [];
+        foreach ($form['prepared'] as $field) {
+            if ($field instanceof BazarField) {
+                $propName = $field->getPropertyName();
+                if (!$field->canRead($data) && !$field->canEdit($data)) {
+                    $restrictedFields[] = $propName;
+                }
+                // TODO when entry update from the form will be done by partiel update ($replace = true), copy the field value when canRead & !canEdit
+            }
+        }
+
+        if (!empty($restrictedFields)) {
+            // if there are some restricted fields, load the previous data by bypassing the rights
+            $previousData = $this->getOne($data['id_fiche'], false, null, false, true);
+
+            // get the value of the restricted fields in the previous data
+            foreach ($restrictedFields as $propName) {
+                $data[$propName] = $previousData[$propName] ?? null;
+            }
+        }
+        return $data;
+    }
+
+    /**
+     * Add the $previousData attributes which match the actual form and which are not in $data
+     * @param array $previousData the data saved in the entry
+     * @param array $data the provided data to update
+     * @return array the data with the merged values
+     * @throws Exception
+     */
+    protected function mergeFields(array $previousData, array $data)
+    {
+        // not possible to init the formManager in the constructor because of circular reference problem
+        $form = $this->wiki->services->get(FormManager::class)->getOne($data['id_typeannonce']);
+
+        foreach ($form['prepared'] as $field) {
+            if ($field instanceof BazarField) {
+                $propName = $field->getPropertyName();
+                if (!isset($data[$propName]) && isset($previousData[$propName])) {
```

Comment:
```suggestion
                if (!empty($propName) && !isset($data[$propName]) && isset($previousData[$propName])) {
```

c'est important pour éviter certains bugs.

Reply:
j'ai fais exprès de ne pas le mettre car si propName est non défini ou est une chaîne vide, la condition est toujours fausse et met un warning.
c'est pour ne pas que l'erreur soit silencieuse, et donc plus facilement détecter de potentiels bugs. Si j'ai bien compris $propName ne devrait avoir ces deux états.

Reply:
Si $propName = '' et que par hasard, il existe déjà $previousData[''] (ce qui peut arriver avec les bases de données jusqu'à maintenant car aucune sécurité à la sauvegarde des données), alors il NE FAUT PAS sauvegarder le champ.
Il ne FAUT PAS générer d'erreur non plus, il FAUT que ce nettoyage de base de données, se fasse silencieusement.
Ça n'est pas une erreur, c'est un nettoyage de la base de données.

Reply:
En effet, j'ai dû ajouter ce test dans BazarField->formatDataBeforeSave pour éviter de sauvegarder les champs spéciaux qui n'ont pas de propertyName, ... ce qui crée de mauvaises données dans la fiche.
Et donc, à cet endroit, si on ne met pas le test, on contourne le fonctionnement de formatDataBeforeSave.

Reply:
je ne parlais pas d'une erreur dans mon cas mais d'un warning qui nous permet à nous dev de se rendre compte que le propertyName a été mal défini (pour les nouveaux champs, ça permet de savoir direct s'il y a un hic)
je suis pas sûr que ça soit super utile de supprimer ces clés vides dans les vieux formulaire mais pourquoi pas. Un commentaire pour ce cas très spécifique serait le bienvenu (je peux m'en charger)

Reply:
En fait, un `empty(propertyName) = true``n'est pas un warning, c'est normal pour labelField, ACLField, HiddenField, ...
Comme ils ne sauvegardent pas de données, ils n'ont pas de propertyname et c'est normal, ....
Je préfère me charger du commentaire.